### DISCUSSION TITLE
Does cross validation (sdmTMB_cv) work with delta models?

### DISCUSSION THREAD

**Discussion by ericward-noaa** [sdmTMB/sdmTMB#111] [Q&A] [answered]
Question via email: does cross validation (sdmTMB_cv) work with delta models?

**Comment by ericward-noaa** [comment]
The delta model features are relatively new, and compatibility with `sdmTMB_cv()` has not yet been added. That said, if you have a large dataset, you can manually do a test/train split or cross validation. For example, with the pcod dataset, we can holdout 10% of the observations as a test set and fit a delta-Gamma model to cod densities

```
data(pcod)
w <- ifelse(runif(nrow(pcod)) < 0.9, 1, 0) # weights, 10% in test set
pcod_spde <- make_mesh(pcod, c("X", "Y"), cutoff = 15)
fit <- sdmTMB(density ~ 1,
              data = pcod, mesh = pcod_spde, spatial = "on",
              spatiotemporal = list("off", "off"), time = "year",
              family = delta_gamma(),
              weights = w # notice weights argument
)
```

Next, we have to make predictions for the binomial component -- and calculate the likelihood
```
pred <- predict(fit, newdata = pcod[which(w==1),])
bin_like <- dbinom(pred$present, size=1, prob=plogis(pred$est1), log=TRUE)
```

The gamma likelihood is a little trickier, 

```
s1 = exp(fit$sd_report$par.fixed[["ln_phi"]])# shape
s2 = pred$est2 / s1 # scale
gamma_like <- dgamma(pred$density, shape = s1, scale = s2, log = TRUE)
```

The total likelihood is just the sum of the 2 components,
`tot_like <- sum(bin_like) + sum(gamma_like[which(is.finite(gamma_like))])`

The 'tot_like' is the total out of sample predictive density, and it's something you want to maximize via model selection. Another issue to be worked out is how to best select the training data. The above approach is totally random, which might be fine. But there's some recent work (see [blockCV](https://cran.r-project.org/web/packages/blockCV/index.html) or [spatialsample](https://cran.r-project.org/web/packages/spatialsample/index.html) R packages) that can do spatial blocking -- to reduce autocorrelation between test / training data. This is a very active area of current research.

**Comment by Raquel-RuizDiaz** [comment]
Good morning,

I was wondering if doing cross-validation with sdmTMB_cv is still not compatible with delta-gamma model. if so, could I manually do a test/train  "leaving the future out" (i.e. leaving the last 3 years of my dataset out and seen how good the model predicts these years) and then using the script you shared here to calculate likelihood of the predictions?

Thanks for your help.
Raquel Ruiz.

**Comment by ericward-noaa** [comment]
Hi -- this is still not implemented for delta models. I think maybe an even easier approach compared to the above script would just be to use the `sdmTMB_cv()` function for each of the 2 model components separately. We have added some arguments that you'll want to pay attention to: 
`lfo` - you'll want to make this TRUE
`lfo_forecast` - how many time steps to forecast (defaults to 1)
`lfo_validations` - how many validation years to use, in your case 3

**Comment by seananderson** [comment, marked as answer]
Cross validation is now working with delta models:

```r
library(sdmTMB)
future::plan(future::multisession)
out_dg <- sdmTMB_cv(
  density ~ depth_scaled,
  data = pcod_2011, 
  mesh = make_mesh(pcod_2011, c("X", "Y"), cutoff = 10),
  family = delta_gamma(), 
  k_folds = 5
)
#> Running fits with `future.apply()`.
#> Set a parallel `future::plan()` to use parallel processing.
out_dg$fold_loglik
#> [1] -531.3995 -702.3048 -591.7539 -608.1418 -611.2388
out_dg$sum_loglik
#> [1] -3044.839
```

**Reply by scottgol** [reply]
sorry, to reply on a two year old comment, but I just want to make sure I understand this correctly: the predictive log-likelihood values from running sdmTMB_cv on a delta model are for the combined model output (so the odds of presence times the predicted abundance/biomass if present), not the predictive log-likelihood values of one of the two parts of the model.  Was that correct?  Thank you

**Reply by seananderson** [reply]
@scottgol sorry I missed replying earlier. Yes, what is returned for delta models is the total predictive log likelihood.

**Comment by James-Thorson-NOAA** [comment]
FWIW, the cv::cv implementation in tinyVAST would also be easy to port over ... it just required implementing to generic functions for package::insights to work.

**Comment by Thom-Teears** [comment]
Hello all,
     I have been doing cross-validation using blockCV to provide spatial predictions in new areas and have been exploring the use of the 'delta_gengamma' distribution for its flexibility. In a previous post above, the likelihood calculation for the gamma distribution was kindly provided. Would it be possible to get advice on how to calculate the likelihood for the generalized gamma distribution as well?
     I am using the 'flexsurv' package and my attempt at calculating the likelihood is below. Thanks for any help you can offer.

  mu.val <- preds$est2
  sigma.val <- sqrt(exp(fit$sd_report$par.fixed[["ln_phi"]]))
  Q.val <- fit$sd_report$par.fixed[["gengamma_Q"]]
  gengamma.like <- dgengamma(x=preds$cpue, mu=mu.val, sigma=sigma.val, Q=Q.val)
  total.like <- sum(bin.like) + sum(gengamma.like[which(is.finite(gengamma.like))])

**Comment by seananderson** [comment]
@Thom-Teears, it gets trickier than that because sdmTMB is parameterizing the generalized gamma in terms of the response scale mean but flexsurv is treating mu as the location parameter in log space. So, you need to do a conversion there first. This function should let you go from sdmTMB parameters to flexsurv parameters:

```r
get_dgmu <- function(Q, sigma, mean) {
  k <- Q^-2
  beta <- Q / sigma
  log_theta <- log(mean) - lgamma((k*beta+1)/beta) + lgamma(k)
  mu <- log_theta + log(k) / beta
  mu
}
```

where `sigma` corresponds directly to `phi` in the sdmTMB output.

So, your code (1) needs to convert the mean predicted value to flexsurv's mu and (2) should be using phi as sigma not sqrt(phi).

However, are you sure you need to calculate this yourself? If using the `sdmTMB_cv()` function, the densities of the left-out data are calculated for you using the same internal gengamma density function used in fitting (and it's *very* easy to mess up the conversion of parameterizations for the gengamma!). You can use blockCV to set up the folds and pass those to `sdmTMB_cv(..., fold_ids = ...)`. If you've done the cross validation yourself with bespoke code, you could still grab the `jnll_obs` element of the left-out data using the same tricks used in `sdmTMB_cv()`. If you need an example of that, let me know.

**Comment by Thom-Teears** [comment]
Thanks for the response and advice. I didn't realize that I could use sdmTMB_cv with spatial blocking so that is very helpful. The only issue I struggle with is how to include vessel ID in the cross-validation. Specifically, for vessels that were in the test dataset that weren't in the training set. I typically have to  remove those vessels from the test data but if I used sdmTMB_cv, I would have to remove them from the entire dataset.

Also, can I get verification that I would not exponentiate the preds$est2 or would I need to exponentiate them since they are on the log scale? Thanks.
